File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    
    genre_set = set(genres)
    genre_cnt = {} # {classic:1450, pop:3100}
    genre_play = {} # {classic:[(500, 0), (600, 1), (800, 3)]}
    for genre in genre_set:
        genre_cnt[genre] = 0
        genre_play[genre] = []
    #print(genre_cnt)
    #print(genre_play)
    
    for i in range(len(genres)):
        genre_cnt[genres[i]] += plays[i]
        genre_play[genres[i]].append((plays[i], i))
    
    #print(genre_cnt)
    #print(genre_play)
    
    ranking = []
    for genre, cnt in genre_cnt.items():
        ranking.append([cnt, genre])
    ranking.sort(reverse = True)
    #print(ranking)
    
    for _, genre in ranking:
        arr = genre_play[genre]
        arr.sort(key = lambda x: (-x[0], x[1]))
        #print(arr)
        for i in range(len(arr)):
            if i >= 2:
                break
            answer.append(arr[i][1])
    
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    /*
    재생수 합 많은 장르 > 한 장르 내 많은 재생수 > 낮은 고유 번호
    한 장르당 최대 2곡
    {장르, 총재생수, [재생수, 고유번호]}
    key:value는 map
    */
    map<string, int> genre_freq;
    map<string, vector<pair<int, int>>> play_freq;
    
    for (int i=0;i<genres.size();i++){
        genre_freq[genres[i]] += plays[i];
        play_freq[genres[i]].push_back({plays[i], i});
    }
    /*
    for (auto& [g, c] : genre_freq){
        cout << g << c << endl;
    }
    */
    
    // genre_freq 정렬
    vector<pair<int, string>> sorted_genre_freq;
    for (auto& [genre, total] : genre_freq){
        sorted_genre_freq.push_back({total, genre});
    }
    sort(sorted_genre_freq.begin(), sorted_genre_freq.end(), [](auto& a, auto& b){
        return a.first > b.first;
    });
    
    // play_freq 정렬(재생수 내림차순, 고유번호 오름차순)
    for (auto& [genre, songs] : play_freq){
        sort(songs.begin(), songs.end(), [](auto& a, auto& b){
            // first는 재생수, second는 고유번호
            if (a.first != b.first) return a.first > b.first;
            return a.second < b.second;
        });
    }
    
    // 한 장르당 최대 2곡
    for (auto& [total, genre] : sorted_genre_freq){
        auto& songs = play_freq[genre];
        for (int i=0; i < 2; i++){
            if (i >= songs.size()) break;
            answer.push_back(songs[i].second);
        }
    }
    
    
    return answer;
}